shared/src/sources/balancer/pool_fetching.rs

```diff
@@ -27,18 +27,138 @@ use std::{
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum BalancerPoolState {
+    Weighted(WeightedPoolState),
+    Stable(TokenState),
+}
+
+impl WeightedPoolState {
+    pub fn balance(&self) -> U256 {
+        self.token_state.balance
+    }
+
+    pub fn scaling_exponent(&self) -> u8 {
+        self.token_state.scaling_exponent
+    }
+}
+
+impl BalancerPoolState {
+    pub fn balance(&self) -> U256 {
+        match self {
+            BalancerPoolState::Weighted(pool) => pool.balance(),
+            BalancerPoolState::Stable(pool) => pool.balance,
+        }
+    }
+
+    pub fn scaling_exponent(&self) -> u8 {
+        match self {
+            BalancerPoolState::Weighted(pool) => pool.scaling_exponent(),
+            BalancerPoolState::Stable(pool) => pool.scaling_exponent,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum BalancerPool {
+    Weighted(WeightedPool),
+    Stable(StablePool),
+}
+
+impl BalancerPool {
+    pub fn pool_id(&self) -> H256 {
+        match self {
+            BalancerPool::Weighted(pool) => pool.pool_id,
+            BalancerPool::Stable(pool) => pool.pool_id,
+        }
+    }
+
+    pub fn paused(&self) -> bool {
+        match self {
+            BalancerPool::Weighted(pool) => pool.paused,
+            BalancerPool::Stable(pool) => pool.paused,
+        }
+    }
+
+    pub fn reserve_keys(&self) -> HashSet<H160> {
+        match self {
+            BalancerPool::Weighted(pool) => pool.reserves.keys().copied().collect(),
+            BalancerPool::Stable(pool) => pool.reserves.keys().copied().collect(),
+        }
+    }
+
+    pub fn swap_fee_percentage(&self) -> Bfp {
+        match self {
+            BalancerPool::Weighted(pool) => pool.swap_fee_percentage,
+            BalancerPool::Stable(pool) => pool.swap_fee_percentage,
+        }
+    }
+
+    pub fn pool_type(&self) -> PoolType {
+        match self {
+            BalancerPool::Weighted(_) => PoolType::Weighted,
+            BalancerPool::Stable(_) => PoolType::Stable,
+        }
+    }
+
+    pub fn reserves(&self) -> HashMap<H160, BalancerPoolState> {
+        match self {
+            BalancerPool::Weighted(pool) => pool
+                .clone()
+                .reserves
+                .into_iter()
+                .map(|(k, v)| (k, BalancerPoolState::Weighted(v)))
+                .collect(),
+            BalancerPool::Stable(pool) => pool
+                .clone()
+                .reserves
+                .into_iter()
+                .map(|(k, v)| (k, BalancerPoolState::Stable(v)))
+                .collect(),
+        }
+    }
+
+    pub fn as_weighted(&self) -> &WeightedPool {
+        if let BalancerPool::Weighted(pool) = self {
+            pool
+        } else {
+            panic!("Not a weighted pool!")
+        }
+    }
+
+    pub fn as_stable(&self) -> &StablePool {
+        if let BalancerPool::Stable(pool) = self {
+            pool
+        } else {
+            panic!("Not a stable pool!")
+        }
+    }
```

Comment:
In a followup PR - I have changed this to `try_as_weighted` and `try_as_stable` returning `Result<_>` which I believe is more appropriate than the panic here.